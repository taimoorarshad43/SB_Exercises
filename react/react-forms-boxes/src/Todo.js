import React, { useState } from "react";

function Todo({ task = "default todo", id = "1", remove, update }) {
  const [editTask, setEditTask] = useState(task);
  const [isEditing, setIsEditing] = useState(false);

  // Update editTask when task prop changes
  React.useEffect(() => {
    setEditTask(task);
  }, [task]);

  const toggleEdit = () => {
    setIsEditing(edit => !edit);
  };

  const handleChange = evt => {
    setEditTask(evt.target.value);
  };

  const handleDelete = () => remove(id);

  const handleUpdate = evt => {
    evt.preventDefault();
    update(id, editTask);
    setIsEditing(false);
  };

  // default todo view
  let jsx = (
    <div style={{ display: 'flex', alignItems: 'center', margin: '5px 0', padding: '10px', border: '1px solid #ccc', borderRadius: '5px' }}>
      <li style={{ flex: 1, margin: 0, listStyle: 'none' }}>{task}</li>
      <button onClick={toggleEdit} style={{ margin: '0 5px', padding: '5px 10px', backgroundColor: '#28a745', color: 'white', border: 'none', borderRadius: '3px', cursor: 'pointer' }}>Edit</button>
      <button onClick={handleDelete} style={{ padding: '5px 10px', backgroundColor: '#dc3545', color: 'white', border: 'none', borderRadius: '3px', cursor: 'pointer' }}>X</button>
    </div>
  );

  // todo view when editing
  if (isEditing) {
    jsx = (
      <div style={{ display: 'flex', alignItems: 'center', margin: '5px 0', padding: '10px', border: '1px solid #ccc', borderRadius: '5px' }}>
        <form onSubmit={handleUpdate} style={{ display: 'flex', flex: 1, alignItems: 'center' }}>
          <input 
            type="text" 
            value={editTask} 
            onChange={handleChange} 
            style={{ flex: 1, padding: '5px', marginRight: '10px' }}
          />
          <button type="submit" style={{ padding: '5px 10px', backgroundColor: '#007bff', color: 'white', border: 'none', borderRadius: '3px', cursor: 'pointer' }}>Update!</button>
        </form>
      </div>
    );
  }

  return jsx;
}

export default Todo;
