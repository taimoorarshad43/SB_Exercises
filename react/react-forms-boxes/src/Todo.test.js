import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Todo from "./Todo";

it("renders without crashing", function() {
  render(<Todo />);
});

it("matches snapshot", function() {
  const { asFragment } = render(<Todo />);
  expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot when editing", function() {
  const { asFragment, getByText } = render(<Todo />);
  const editButton = getByText("Edit");
  fireEvent.click(editButton);
  expect(asFragment()).toMatchSnapshot();
});

it("runs the update function on form submit", function() {
  const updateMock = jest.fn();
  const { getByText } = render(<Todo update={updateMock} />);
  const editButton = getByText("Edit");
  fireEvent.click(editButton);
  const updateButton = getByText("Update!");
  fireEvent.click(updateButton);
  expect(updateMock).toHaveBeenCalled();
});

it("runs the delete function on button click", function() {
  const removeMock = jest.fn();
  const { getByText } = render(<Todo remove={removeMock} />);
  const deleteButton = getByText("X");
  fireEvent.click(deleteButton);
  expect(removeMock).toHaveBeenCalled();
});

it("toggles between edit and view modes", function() {
  const { getByText, queryByText } = render(<Todo task="Test task" />);
  
  // Initially in view mode
  expect(getByText("Test task")).toBeInTheDocument();
  expect(queryByText("Update!")).not.toBeInTheDocument();
  
  // Click edit button
  const editButton = getByText("Edit");
  fireEvent.click(editButton);
  
  // Now in edit mode
  expect(queryByText("Test task")).not.toBeInTheDocument();
  expect(getByText("Update!")).toBeInTheDocument();
});

it("updates task text when editing", function() {
  const updateMock = jest.fn();
  const { getByText, getByDisplayValue } = render(<Todo task="Original task" update={updateMock} />);
  
  const editButton = getByText("Edit");
  fireEvent.click(editButton);
  
  const input = getByDisplayValue("Original task");
  fireEvent.change(input, { target: { value: "Updated task" } });
  
  const updateButton = getByText("Update!");
  fireEvent.click(updateButton);
  
  expect(updateMock).toHaveBeenCalledWith(expect.any(String), "Updated task");
});
