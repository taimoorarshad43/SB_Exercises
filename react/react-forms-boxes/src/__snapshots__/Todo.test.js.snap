// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: flex; align-items: center; margin: 5px 0px; padding: 10px; border: 1px solid #ccc; border-radius: 5px;"
  >
    <li
      style="flex: 1; margin: 0px; list-style: none;"
    >
      default todo
    </li>
    <button
      style="margin: 0px 5px; padding: 5px 10px; background-color: rgb(40, 167, 69); color: white; border-radius: 3px; cursor: pointer;"
    >
      Edit
    </button>
    <button
      style="padding: 5px 10px; background-color: rgb(220, 53, 69); color: white; border-radius: 3px; cursor: pointer;"
    >
      X
    </button>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot when editing 1`] = `
<DocumentFragment>
  <div
    style="display: flex; align-items: center; margin: 5px 0px; padding: 10px; border: 1px solid #ccc; border-radius: 5px;"
  >
    <form
      style="display: flex; flex: 1; align-items: center;"
    >
      <input
        style="flex: 1; padding: 5px; margin-right: 10px;"
        type="text"
        value="default todo"
      />
      <button
        style="padding: 5px 10px; background-color: rgb(0, 123, 255); color: white; border-radius: 3px; cursor: pointer;"
        type="submit"
      >
        Update!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
